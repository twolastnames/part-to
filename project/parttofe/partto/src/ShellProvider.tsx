import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "./App.scss";

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

export enum Theme {
  Default = "default",
  Dark = "dark",
}

export function changeTheme(theme: Theme) {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeChangeRef?.current && themeChangeRef.current(theme);
  themeNode.className = `${theme}Theme`;
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(Theme.Default);
  themeChangeRef = useRef(setTheme);
  return (
    <div id="theme" className="defaultTheme">
      <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
    </div>
  );
};
