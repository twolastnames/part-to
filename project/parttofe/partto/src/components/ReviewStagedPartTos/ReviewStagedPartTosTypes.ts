import { MutableRefObject } from "react";
import { RunStateId, TaskDefinitionId } from "../../api/sharedschemas";

export interface ReviewStagedPartTosProps {
  taskDefinitions: Array<TaskDefinitionId>;
  runState: MutableRefObject<RunStateId>;
}
