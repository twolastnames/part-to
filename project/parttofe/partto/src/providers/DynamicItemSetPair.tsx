import React, {
  createContext,
  useState,
  Context,
  MutableRefObject,
  createRef,
  useContext,
  SetStateAction,
} from "react";
import { PropsWithChildren } from "react";
import { Duration, getDuration } from "../shared/duration";
import { debounce, DebouncedFunc } from "lodash";

export type DynamicItemSetPairType = {
  selected: number;
  setSelected: (selected: number) => void;
  getTotal: () => number;
  goBack: () => void;
  goForward: () => void;
  showDuration: Duration;
  togglePause: () => void;
  paused: boolean;
  nextShowDuration: () => void;
};

export const UndefinedDynamicItemSetPair = {
  selected: 0,
  setSelected: (_: number) => undefined,
  goBack: () => undefined,
  goForward: () => undefined,
  getTotal: () => 0,
  showDuration: getDuration(0),
  togglePause: () => undefined,
  paused: true,
  nextShowDuration: () => undefined,
};

export function useSingleOfPair({ context }: ContextDescription) {
  return useContext(context) || UndefinedDynamicItemSetPair;
}

type ContextType = Context<undefined | DynamicItemSetPairType>;

export type ContextDescription = {
  context: ContextType;
  setCount: (value: number) => void;
  setSettingKey: (settingKey: string) => void;
  setDurations: (value: Array<Duration>) => void;
};

const defaultDurations = [1000, 2000, 4000, 8000];

const getDurations = (key: string) =>
  JSON.parse(
    localStorage.getItem(`${key}Durations`) || JSON.stringify(defaultDurations),
  ).map(getDuration);

const setDurations = (key: string, value: Array<Duration>) => {
  localStorage.setItem(
    `${key}Durations`,
    JSON.stringify(value.map((duration) => duration.toMilliseconds())),
  );
};

const leftCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const leftSettingKey: MutableRefObject<string | null> = createRef();
const getLeftSettingKey = () => leftSettingKey.current || "left";
function setLeftCount(value: number) {
  leftCount.current && leftCount.current(value);
}
const LeftDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const LeftContext: ContextDescription = {
  context: LeftDynamicItemSetPairContext,
  setCount: setLeftCount,
  setSettingKey: (value: string) => {
    leftSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getLeftSettingKey(), value);
  },
};

const rightCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const rightSettingKey: MutableRefObject<string | null> = createRef();
const getRightSettingKey = () => leftSettingKey.current || "left";

function setRightCount(value: number) {
  rightCount.current && rightCount.current(value);
}
const RightDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const RightContext: ContextDescription = {
  context: RightDynamicItemSetPairContext,
  setCount: setRightCount,
  setSettingKey: (value: string) => {
    rightSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getRightSettingKey(), value);
  },
};

type GoADirection = (
  totalPages: number,
  setSelected: (mutator: (arg: ProviderState) => ProviderState) => void,
) => void;

const goBack: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected <= 0 ? length - 1 : last.selected - 1,
  }));
};

const goForward: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected >= length - 1 ? 0 : last.selected + 1,
  }));
};

type ProviderState = {
  selected: number;
  showDuration: number;
  paused: boolean;
  count: number;
};

function useProvider({
  setCount,
  getSettingKey,
  togglePauseClickSlower,
}: {
  getSettingKey: () => string;
  setCount: MutableRefObject<((value: number) => void) | null>;
  togglePauseClickSlower: DebouncedFunc<
    (setState: (arg: SetStateAction<ProviderState>) => void) => void
  >;
}) {
  const durations = getDurations(getSettingKey());
  const [{ count, selected, showDuration, paused }, setState] =
    useState<ProviderState>({
      selected: 0,
      showDuration: Math.floor(durations.length / 2),
      paused: false,
      count: 0,
    });

  setCount.current = (value: number) => {
    if (value === count) {
      return;
    }
    setState((previous) => ({ ...previous, count: value }));
  };

  if (selected !== 0 && (selected < 0 || selected >= count)) {
    setState((current) => ({ ...current, selected: 0 }));
  }

  const duration = durations[showDuration];
  return {
    selected: selected,
    setSelected: (selected: number) => {
      setState((current) => ({ ...current, selected }));
    },
    goBack: () => {
      goBack(count, setState);
    },
    goForward: () => {
      goForward(count, setState);
    },
    showDuration: duration,
    paused: paused,
    getTotal: () => count,
    togglePause: () => {
      togglePauseClickSlower(setState);
    },
    nextShowDuration: () => {
      const next = showDuration + 1;
      setState((previous) => ({
        ...previous,
        showDuration: next >= durations.length ? 0 : next,
      }));
    },
  };
}

const leftTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function LeftDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <LeftDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: leftCount,
        getSettingKey: getLeftSettingKey,
        togglePauseClickSlower: leftTogglePauseClickSlower,
      })}
    >
      {children}
    </LeftDynamicItemSetPairContext.Provider>
  );
}

const rightTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function RightDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <RightDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: rightCount,
        getSettingKey: getRightSettingKey,
        togglePauseClickSlower: rightTogglePauseClickSlower,
      })}
    >
      {children}
    </RightDynamicItemSetPairContext.Provider>
  );
}
