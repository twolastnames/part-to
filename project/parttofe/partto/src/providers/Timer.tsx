import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { PropsWithChildren } from "react";
import { DateTime, getDateTime } from "../shared/dateTime";
import { Duration, getDuration } from "../shared/duration";
import { RunState, TaskDefinitionId } from "../api/sharedschemas";
import { Timer } from "../components/Timer/Timer";
import {
  addAlarmNote,
  removeNote,
} from "../components/Layout/NavigationBar/Noted/Noted";
import {
  addRunStateListener,
  removeRunStateListener,
} from "../shared/runStateMessage";

type TimerDescription = {
  component: ReactNode;
  duration: Duration;
  started: DateTime;
  message: string;
  enforced: boolean;
  task: TaskDefinitionId;
  setOnLocate: (callback: () => void) => void;
};

export type TimerType = {
  timers: { [id: string]: TimerDescription };
};

export const TimerContext = createContext<undefined | TimerType>(undefined);

type TaskKeyed = { task: TaskDefinitionId };

type TimerFactory<T> = (args: T) => {
  component: ReactNode;
  duration: Duration;
  task: TaskDefinitionId;
};

function getMappedTimers<T extends TaskKeyed>(
  getTimer: TimerFactory<T>,
  timers: Array<T> | undefined,
): { [id: string]: TimerDescription } {
  return (
    timers?.reduce(
      (current, timer) => ({
        ...current,
        [timer.task]: getTimer(timer),
      }),
      {},
    ) || ({} as { [id: TaskDefinitionId]: T })
  );
}

export function TimerProvider({ children }: PropsWithChildren) {
  const [runState, setRunState] = useState<RunState | undefined>();
  const offsets = useRef<{
    [id: string]: Duration;
  }>({});
  const onLocates = useRef<{ [task: string]: () => void }>({});
  const audio = useRef(new Audio(require("./messageAlert.mp3")));
  const timestamp = runState?.timestamp;
  const timers = useMemo(
    () => ({
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: true,
          message: "Duty Complete?",
          component: (
            <Timer
              start={started}
              duration={duration}
              adjustment={{
                offset: offsets.current[task] || getDuration(0),
                setOffset: (value: Duration) => (offsets.current[task] = value),
              }}
            />
          ),
        }),
        runState?.timers.enforced || [],
      ),
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: false,
          message: "Past Expected Time",
          component: <Timer start={started} duration={duration} />,
        }),
        runState?.timers.laxed || [],
      ),
      ...getMappedTimers(
        ({ task, till }) => ({
          task,
          started: timestamp,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration: till,
          enforced: true,
          message: "Need to Start Duty",
          component: <Timer start={timestamp} duration={till} />,
        }),
        runState?.timers.imminent || [],
      ),
    }),
    [runState, offsets, timestamp],
  );

  useEffect(() => {
    addRunStateListener(setRunState);
    return () => {
      removeRunStateListener(setRunState);
    };
  }, [setRunState]);

  useEffect(() => {
    const id = setInterval(() => {
      const now = getDateTime();
      for (const [key, timer] of Object.entries(timers)) {
        const expireAt = timer.started
          .add(offsets.current[key] || getDuration(0))
          .add(timer.duration);
        if (!timer.enforced || now.sinceEpoch() < expireAt.sinceEpoch()) {
          continue;
        }
        addAlarmNote({
          key,
          heading: "Alarm Expired",
          detail: timer.message,
          onClick: () => {
            onLocates.current[key]?.();
          },
        });
        if (!audio.current.paused) {
          return;
        }
        try {
          audio.current.play();
        } catch (NotAllowedError) {
          console.error(
            "could not start an alarm... likely because the page has not been interacted with",
          );
        }
        audio.current.loop = true;
        return;
      }
      for (const key of Object.keys(timers)) {
        removeNote(key);
      }
      audio.current.pause();
    }, 1000);
    return () => {
      clearInterval(id);
    };
  }, [offsets, runState, timers]);

  return (
    <TimerContext.Provider
      value={{
        timers,
      }}
    >
      {children}
    </TimerContext.Provider>
  );
}

export function useTimerProvider({
  task,
  onLocate,
}: {
  task: TaskDefinitionId;
  onLocate?: () => void;
}) {
  const timers = useContext(TimerContext)?.timers;
  onLocate && timers?.[task]?.setOnLocate(onLocate);
  return timers?.[task]?.component || <></>;
}
