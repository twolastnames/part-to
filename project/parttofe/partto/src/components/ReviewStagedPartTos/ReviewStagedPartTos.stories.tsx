import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { ReviewStagedPartTos } from "./ReviewStagedPartTos";
import { ShellProvider } from "../../providers/ShellProvider";
import { MemoryRouter } from "react-router-dom";

const meta: Meta<typeof ReviewStagedPartTos> = {
  component: ReviewStagedPartTos,
  decorators: (Story) => (
    <ShellProvider>
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof ReviewStagedPartTos>;

export const Simple: Story = {
  args: {
    taskDefinitions: [],
  },
};
