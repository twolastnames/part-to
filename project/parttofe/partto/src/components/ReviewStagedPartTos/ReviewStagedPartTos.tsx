import React from "react";

import { RunStateId, TaskDefinitionId } from "../../api/sharedschemas";
import { DynamicItemSet } from "../DynamicItemSet/DynamicItemSet";
import { Spinner } from "../Spinner/Spinner";
import { Cancel, Oven } from "../Icon/Icon";
import { doRunvoidPost } from "../../api/runvoidpost";
import { useNavigate } from "react-router-dom";
import { getRoute } from "../../routes";
import { doRunstartPost } from "../../api/runstartpost";
import { ReviewStagedPartTosProps } from "./ReviewStagedPartTosTypes";
import { LeftContext, RightContext } from "../../providers/DynamicItemSetPair";
import { useRunState } from "../../hooks/runState";
import {
  StagedClassNames,
  TaskDefinition,
} from "../TaskDefinition/TaskDefinition";

export function ReviewStagedPartTosIdFromer() {
  const response = useRunState();
  return (
    <Spinner responses={[response]}>
      <ReviewStagedPartTos taskDefinitions={response?.data?.staged || []} />
    </Spinner>
  );
}

function getItems(
  navigate: (arg: string) => void,
  runState: RunStateId,
  taskDefinitions: Array<TaskDefinitionId>,
) {
  return taskDefinitions.map(
    (taskDefinitionId: TaskDefinitionId, index: number) => ({
      key: taskDefinitionId,
      listView: <>{taskDefinitionId}</>,
      detailView: (
        <TaskDefinition
          task={taskDefinitionId}
          runState={runState}
          locatable={{
            onLocate: (setter) => () => {
              setter(index);
            },
            context: RightContext,
          }}
          classNames={StagedClassNames}
        />
      ),
      itemOperations: [
        {
          text: "Skip and Void",
          icon: Cancel,
          onClick: () => {
            doRunvoidPost({
              body: {
                runState: runState,
                definitions: [taskDefinitionId],
              },
              on200: ({ runState }) => {
                navigate(getRoute("StageMeal", { runState: runState }));
              },
            });
          },
        },
      ],
    }),
  );
}

export function ReviewStagedPartTos({
  taskDefinitions,
}: ReviewStagedPartTosProps) {
  const navigate = useNavigate();
  const response = useRunState();
  const items = getItems(
    navigate,
    response.data?.runState as RunStateId,
    taskDefinitions,
  );
  return (
    <Spinner responses={[response]}>
      <DynamicItemSet
        items={items}
        context={RightContext}
        setOperations={[
          {
            text: "Start Cooking",
            icon: Oven,
            onClick: () => {
              doRunstartPost({
                body: { runState: response.data?.runState },
                on200: ({ runState }) => {
                  LeftContext.setCount(0);
                  LeftContext.setSettingKey("duties");
                  RightContext.setCount(0);
                  RightContext.setSettingKey("tasks");
                  navigate(getRoute("CookMeal", { runState }));
                },
              });
            },
          },
        ]}
        emptyPage={
          <>
            {[
              "Select recipes from the first pane with the plus",
              "and add tasks for meal a meal to the second",
              "pane. When all your meal tasks are in the second pane,",
              "click the oven that will appear to start cooking.",
            ].join(" ")}
          </>
        }
      />
    </Spinner>
  );
}
