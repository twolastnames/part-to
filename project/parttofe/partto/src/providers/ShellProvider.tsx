import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
