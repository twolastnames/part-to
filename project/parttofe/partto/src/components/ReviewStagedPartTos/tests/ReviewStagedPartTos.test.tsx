import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../providers/ShellProvider";
import { ReviewStagedPartTos } from "../ReviewStagedPartTos";

test("snapshot", () => {
  render(
    <ShellProvider>
      <ReviewStagedPartTos runState="Hello" />
    </ShellProvider>,
  );
  const component = screen.getByTestId("DynamicItemSet");
  expect(component).toMatchSnapshot();
});
