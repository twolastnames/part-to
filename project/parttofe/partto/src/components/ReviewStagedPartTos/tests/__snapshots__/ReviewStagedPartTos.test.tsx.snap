// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="dynamicItemSet"
  data-testid="DynamicItemSet"
>
  <div
    class="buttonSet"
    data-testid="ButtonSet"
  >
    <div
      class="container"
    >
      <button
        aria-label="Start Cooking"
        class="mantine-focus-auto mantine-active button m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
        data-testid="Button"
        style="--ai-color: var(--mantine-color-white);"
        type="button"
      >
        <span
          class="m_8d3afb97 mantine-ActionIcon-icon"
        >
          <svg
            class="tabler-icon tabler-icon-cooker "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 7h.01"
            />
            <path
              d="M15 7h.01"
            />
            <path
              d="M9 7h.01"
            />
            <path
              d="M5 3m0 2a2 2 0 0 1 2 -2h10a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-10a2 2 0 0 1 -2 -2z"
            />
            <path
              d="M9 15h6"
            />
            <path
              d="M5 11h14"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="multiple"
    data-testid="Multiple"
  >
    <div
      class="switchButton"
    >
      <div
        class="container"
      >
        <button
          aria-label="See all 2 items in a list"
          class="mantine-focus-auto mantine-active button m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
          data-testid="Button"
          style="--ai-color: var(--mantine-color-white);"
          type="button"
        >
          <span
            class="m_8d3afb97 mantine-ActionIcon-icon"
          >
            <svg
              class="tabler-icon tabler-icon-list "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 6l11 0"
              />
              <path
                d="M9 12l11 0"
              />
              <path
                d="M9 18l11 0"
              />
              <path
                d="M5 6l0 .01"
              />
              <path
                d="M5 12l0 .01"
              />
              <path
                d="M5 18l0 .01"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <span
      data-testid="Detail"
    >
      <div
        class="multiple"
        data-testid="Carousel"
      >
        <div
          class="carousel"
        >
          <div
            class="carouselButton"
          >
            <div
              class="container"
            >
              <button
                aria-label="Back One"
                class="mantine-focus-auto mantine-active button m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
                data-testid="Button"
                style="--ai-color: var(--mantine-color-white);"
                type="button"
              >
                <span
                  class="m_8d3afb97 mantine-ActionIcon-icon"
                >
                  <svg
                    class="tabler-icon tabler-icon-caret-up "
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M18 14l-6 -6l-6 6h12"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="listView"
            />
          </div>
          <div
            class="carouselButton"
          >
            <div
              class="container"
            >
              <button
                aria-label="Forward One"
                class="mantine-focus-auto mantine-active button m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
                data-testid="Button"
                style="--ai-color: var(--mantine-color-white);"
                type="button"
              >
                <span
                  class="m_8d3afb97 mantine-ActionIcon-icon"
                >
                  <svg
                    class="tabler-icon tabler-icon-caret-down "
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 10l6 6l6 -6h-12"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          class="iterationProgress"
          data-testid="IterationProgress"
        >
          <span>
            <span
              data-testid="Timer"
            >
              <span
                data-testid="LocalProgress"
              >
                <div
                  class="m_b32e4812 mantine-RingProgress-root"
                  data-testid="Progress"
                  style="--rp-size: calc(3.5rem * var(--mantine-scale)); --rp-label-offset: calc(1rem * var(--mantine-scale));"
                  title="Unpaused"
                >
                  <svg
                    class="m_d43b5134 mantine-RingProgress-svg"
                  >
                    <circle
                      class="m_b1ca1fbf mantine-RingProgress-curve"
                      cx="28"
                      cy="28"
                      fill="none"
                      r="17.2"
                      stroke-dasharray="108.043769586668, 0.027017696820872225"
                      stroke-dashoffset="0"
                      stroke-linecap="butt"
                      stroke-width="8"
                      transform="scale(1, -1) translate(0, -56)"
                    />
                    <circle
                      class="m_b1ca1fbf mantine-RingProgress-curve"
                      cx="28"
                      cy="28"
                      fill="none"
                      r="17.2"
                      stroke-dasharray="0.027017696820872225, 108.043769586668"
                      stroke-dashoffset="108.07078728348888"
                      stroke-linecap="butt"
                      stroke-width="8"
                      style="--curve-color: var(--secondary-color);"
                    />
                  </svg>
                  <div
                    class="m_b23f9dc4 mantine-RingProgress-label"
                  >
                    <div
                      class="label"
                    >
                      1
                    </div>
                  </div>
                </div>
              </span>
            </span>
          </span>
          <span
            class="count"
          >
            <span
              data-testid="Count"
            >
              <span
                data-testid="LocalProgress"
              >
                <div
                  class="m_b32e4812 mantine-RingProgress-root"
                  data-testid="Progress"
                  style="--rp-size: calc(3.5rem * var(--mantine-scale)); --rp-label-offset: calc(1rem * var(--mantine-scale));"
                  title="4 seconds"
                >
                  <svg
                    class="m_d43b5134 mantine-RingProgress-svg"
                  >
                    <circle
                      class="m_b1ca1fbf mantine-RingProgress-curve"
                      cx="28"
                      cy="28"
                      fill="none"
                      r="17.2"
                      stroke-dasharray="54.03539364174444, 54.03539364174444"
                      stroke-dashoffset="0"
                      stroke-linecap="butt"
                      stroke-width="8"
                      transform="scale(1, -1) translate(0, -56)"
                    />
                    <circle
                      class="m_b1ca1fbf mantine-RingProgress-curve"
                      cx="28"
                      cy="28"
                      fill="none"
                      r="17.2"
                      stroke-dasharray="54.03539364174444, 54.03539364174444"
                      stroke-dashoffset="108.07078728348888"
                      stroke-linecap="butt"
                      stroke-width="8"
                      style="--curve-color: var(--secondary-color);"
                    />
                  </svg>
                  <div
                    class="m_b23f9dc4 mantine-RingProgress-label"
                  >
                    <div
                      class="label"
                    >
                      2
                    </div>
                  </div>
                </div>
              </span>
            </span>
          </span>
        </div>
        <div
          class="buttonSet"
        >
          <div
            class="buttons"
          >
            <div
              class="buttonSet"
              data-testid="ButtonSet"
            >
              <div
                class="container"
              >
                <button
                  aria-label="Skip and Void"
                  class="mantine-focus-auto mantine-active button m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
                  data-testid="Button"
                  style="--ai-color: var(--mantine-color-white);"
                  type="button"
                >
                  <span
                    class="m_8d3afb97 mantine-ActionIcon-icon"
                  >
                    <svg
                      class="tabler-icon tabler-icon-cancel "
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 12a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
                      />
                      <path
                        d="M18.364 5.636l-12.728 12.728"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>
</div>
`;
